File: packages/node/iii/src/metrics.ts
import * as os from 'os'
import * as fs from 'fs'
import * as path from 'path'
import type {
  WorkerMetrics,
  ProcessMetrics,
  PerformanceMetrics,
  ExtendedMetrics,
  KubernetesIdentifiers,
  KubernetesCoreMetrics,
  KubernetesResourceMetrics,
  KubernetesExtendedMetrics,
} from './bridge-types'

let lastCpuUsage: NodeJS.CpuUsage | null = null
let lastCpuTime: [number, number] | null = null
let lastNetworkStats: { rx: number; tx: number; time: number } | null = null
let lastDiskStats: { read: number; write: number; time: number } | null = null

let invocationCount = 0
let totalLatencyMs = 0
let errorCount = 0
let lastInvocationWindow = Date.now()

const processStartTime = Date.now()

export function recordInvocation(latencyMs: number, isError: boolean = false): void {
  invocationCount++
  totalLatencyMs += latencyMs
  if (isError) {
    errorCount++
  }
}

function getCpuPercent(): number {
  const currentUsage = process.cpuUsage(lastCpuUsage ?? undefined)
  const currentTime = process.hrtime()

  if (lastCpuUsage === null || lastCpuTime === null) {
    lastCpuUsage = process.cpuUsage()
    lastCpuTime = process.hrtime()
    return 0
  }

  const elapsedTime = (currentTime[0] - lastCpuTime[0]) * 1e6 + (currentTime[1] - lastCpuTime[1]) / 1000

  if (elapsedTime <= 0) {
    return 0
  }

  const cpuTimeUsed = currentUsage.user + currentUsage.system
  const cpuPercent = (cpuTimeUsed / elapsedTime) * 100

  lastCpuUsage = process.cpuUsage()
  lastCpuTime = process.hrtime()

  return Math.min(cpuPercent, 100 * os.cpus().length)
}

function getMemoryMetrics(): { used: number; total: number } {
  const memUsage = process.memoryUsage()
  return {
    used: memUsage.rss,
    total: os.totalmem(),
  }
}

function getThreadCount(): number {
  const uvThreadPoolSize = parseInt(process.env.UV_THREADPOOL_SIZE || '4', 10)
  return 1 + uvThreadPoolSize
}

function isKubernetes(): boolean {
  return !!process.env.KUBERNETES_SERVICE_HOST
}

function getKubernetesIdentifiers(): KubernetesIdentifiers | undefined {
  if (!isKubernetes()) {
    return undefined
  }

  return {
    cluster: process.env.CLUSTER_NAME,
    namespace: process.env.POD_NAMESPACE || readK8sFile('namespace'),
    pod_name: process.env.POD_NAME || process.env.HOSTNAME,
    container_name: process.env.CONTAINER_NAME,
    node_name: process.env.NODE_NAME,
    pod_uid: process.env.POD_UID || readK8sFile('pod-uid'),
  }
}

function readK8sFile(filename: string): string | undefined {
  const basePath = '/var/run/secrets/kubernetes.io/serviceaccount'
  const filePath = path.join(basePath, filename)
  try {
    if (fs.existsSync(filePath)) {
      return fs.readFileSync(filePath, 'utf-8').trim()
    }
  } catch {
  }
  return undefined
}

function readCgroupValue(v1Path: string, v2Path: string): number | undefined {
  const paths = [v2Path, v1Path]
  for (const p of paths) {
    try {
      if (fs.existsSync(p)) {
        const content = fs.readFileSync(p, 'utf-8').trim()
        const value = parseInt(content, 10)
        if (!isNaN(value) && value !== -1) {
          return value
        }
      }
    } catch {
    }
  }
  return undefined
}

function getCgroupCpuUsageCores(): number | undefined {
  try {
    const v2Path = '/sys/fs/cgroup/cpu.stat'
    if (fs.existsSync(v2Path)) {
      const content = fs.readFileSync(v2Path, 'utf-8')
      const match = content.match(/usage_usec\s+(\d+)/)
      if (match) {
        const usec = parseInt(match[1], 10)
        return usec / 1_000_000
      }
    }

    const v1Path = '/sys/fs/cgroup/cpu/cpuacct.usage'
    if (fs.existsSync(v1Path)) {
      const nsec = parseInt(fs.readFileSync(v1Path, 'utf-8').trim(), 10)
      return nsec / 1_000_000_000
    }
  } catch {
  }
  return undefined
}

function getCgroupMemoryWorkingSet(): number | undefined {
  return readCgroupValue(
    '/sys/fs/cgroup/memory/memory.usage_in_bytes',
    '/sys/fs/cgroup/memory.current'
  )
}

function getCgroupMemoryLimit(): number | undefined {
  const value = readCgroupValue(
    '/sys/fs/cgroup/memory/memory.limit_in_bytes',
    '/sys/fs/cgroup/memory.max'
  )
  if (value && value < Number.MAX_SAFE_INTEGER) {
    return value
  }
  return undefined
}

function getCgroupCpuLimitCores(): number | undefined {
  try {
    const v2Path = '/sys/fs/cgroup/cpu.max'
    if (fs.existsSync(v2Path)) {
      const content = fs.readFileSync(v2Path, 'utf-8').trim()
      if (content !== 'max') {
        const [quota, period] = content.split(' ').map((s) => parseInt(s, 10))
        if (quota > 0 && period > 0) {
          return quota / period
        }
      }
    }

    const quotaPath = '/sys/fs/cgroup/cpu/cpu.cfs_quota_us'
    const periodPath = '/sys/fs/cgroup/cpu/cpu.cfs_period_us'
    if (fs.existsSync(quotaPath) && fs.existsSync(periodPath)) {
      const quota = parseInt(fs.readFileSync(quotaPath, 'utf-8').trim(), 10)
      const period = parseInt(fs.readFileSync(periodPath, 'utf-8').trim(), 10)
      if (quota > 0 && period > 0) {
        return quota / period
      }
    }
  } catch {
  }
  return undefined
}

function getCgroupCpuThrottledSeconds(): number | undefined {
  try {
    const v2Path = '/sys/fs/cgroup/cpu.stat'
    if (fs.existsSync(v2Path)) {
      const content = fs.readFileSync(v2Path, 'utf-8')
      const match = content.match(/throttled_usec\s+(\d+)/)
      if (match) {
        return parseInt(match[1], 10) / 1_000_000
      }
    }

    const v1Path = '/sys/fs/cgroup/cpu/cpu.stat'
    if (fs.existsSync(v1Path)) {
      const content = fs.readFileSync(v1Path, 'utf-8')
      const match = content.match(/throttled_time\s+(\d+)/)
      if (match) {
        return parseInt(match[1], 10) / 1_000_000_000
      }
    }
  } catch {
  }
  return undefined
}

function getFilesystemUsage(): number | undefined {
  if (process.platform !== 'linux' && process.platform !== 'darwin') {
    return undefined
  }

  try {
    const { execSync } = require('child_process')
    const output = execSync('df -B1 / 2>/dev/null', { encoding: 'utf-8', timeout: 1000 })
    const lines = output.trim().split('\n')
    if (lines.length >= 2) {
      const parts = lines[1].split(/\s+/)
      if (parts.length >= 4) {
        const used = parseInt(parts[2], 10)
        if (!isNaN(used)) {
          return used
        }
      }
    }
  } catch {
  }
  return undefined
}

function getNodePressure(): {
  memory: boolean | undefined
  disk: boolean | undefined
  pid: boolean | undefined
} {
  const result = { memory: undefined as boolean | undefined, disk: undefined as boolean | undefined, pid: undefined as boolean | undefined }

  if (process.platform !== 'linux') {
    return result
  }

  const extractAvg10 = (line: string): number | undefined => {
    const match = line.match(/avg10=(\d+\.?\d*)/)
    return match ? parseFloat(match[1]) : undefined
  }

  try {
    const content = fs.readFileSync('/sys/fs/cgroup/memory.pressure', 'utf-8')
    for (const line of content.split('\n')) {
      if (line.startsWith('some')) {
        const avg10 = extractAvg10(line)
        if (avg10 !== undefined) {
          result.memory = avg10 > 10.0
        }
      }
    }
  } catch {
  }

  try {
    const content = fs.readFileSync('/sys/fs/cgroup/io.pressure', 'utf-8')
    for (const line of content.split('\n')) {
      if (line.startsWith('some')) {
        const avg10 = extractAvg10(line)
        if (avg10 !== undefined) {
          result.disk = avg10 > 10.0
        }
      }
    }
  } catch {
  }

  try {
    const content = fs.readFileSync('/sys/fs/cgroup/cpu.pressure', 'utf-8')
    for (const line of content.split('\n')) {
      if (line.startsWith('some')) {
        const avg10 = extractAvg10(line)
        if (avg10 !== undefined) {
          result.pid = avg10 > 10.0
        }
      }
    }
  } catch {
  }

  return result
}

function getKubernetesCoreMetrics(): KubernetesCoreMetrics | undefined {
  if (!isKubernetes()) {
    return undefined
  }

  const metrics: KubernetesCoreMetrics = {
    uptime_seconds: Math.floor((Date.now() - processStartTime) / 1000),
    pod_phase: 'Running',
    pod_ready: true,
    cpu_usage_cores: getCgroupCpuUsageCores(),
    memory_working_set_bytes: getCgroupMemoryWorkingSet(),
  }

  const restartsEnv = process.env.CONTAINER_RESTARTS
  if (restartsEnv) {
    metrics.container_restarts_total = parseInt(restartsEnv, 10)
  }

  metrics.last_termination_reason = process.env.LAST_TERMINATION_REASON

  return metrics
}

function getKubernetesResourceMetrics(): KubernetesResourceMetrics | undefined {
  if (!isKubernetes()) {
    return undefined
  }

  const metrics: KubernetesResourceMetrics = {}

  metrics.cpu_limits_cores = getCgroupCpuLimitCores()
  metrics.memory_limits_bytes = getCgroupMemoryLimit()
  metrics.cpu_throttled_seconds_total = getCgroupCpuThrottledSeconds()

  const cpuRequest = process.env.CPU_REQUEST
  if (cpuRequest) {
    if (cpuRequest.endsWith('m')) {
      metrics.cpu_requests_cores = parseInt(cpuRequest, 10) / 1000
    } else {
      metrics.cpu_requests_cores = parseFloat(cpuRequest)
    }
  }

  const memRequest = process.env.MEMORY_REQUEST
  if (memRequest) {
    metrics.memory_requests_bytes = parseK8sMemory(memRequest)
  }

  return Object.keys(metrics).length > 0 ? metrics : undefined
}

function parseK8sMemory(value: string): number | undefined {
  const match = value.match(/^(\d+(?:\.\d+)?)(Ki|Mi|Gi|Ti|Pi|Ei|K|M|G|T|P|E)?$/)
  if (!match) return undefined

  const num = parseFloat(match[1])
  const unit = match[2] || ''

  const multipliers: Record<string, number> = {
    '': 1,
    K: 1000,
    M: 1000 ** 2,
    G: 1000 ** 3,
    T: 1000 ** 4,
    P: 1000 ** 5,
    E: 1000 ** 6,
    Ki: 1024,
    Mi: 1024 ** 2,
    Gi: 1024 ** 3,
    Ti: 1024 ** 4,
    Pi: 1024 ** 5,
    Ei: 1024 ** 6,
  }

  return num * (multipliers[unit] || 1)
}

function getKubernetesExtendedMetrics(): KubernetesExtendedMetrics | undefined {
  if (!isKubernetes()) {
    return undefined
  }

  const metrics: KubernetesExtendedMetrics = {}

  const netStats = getNetworkStats()
  if (netStats) {
    metrics.network_rx_bytes_total = netStats.rx
    metrics.network_tx_bytes_total = netStats.tx
  }

  metrics.fs_usage_bytes = getFilesystemUsage()

  const pressure = getNodePressure()
  metrics.node_memory_pressure = pressure.memory
  metrics.node_disk_pressure = pressure.disk
  metrics.node_pid_pressure = pressure.pid

  return Object.keys(metrics).some((k) => metrics[k as keyof KubernetesExtendedMetrics] !== undefined)
    ? metrics
    : undefined
}

function getNetworkStats(): { rx: number; tx: number } | undefined {
  if (process.platform !== 'linux') {
    return undefined
  }

  try {
    const netDev = fs.readFileSync('/proc/net/dev', 'utf-8')
    const lines = netDev.split('\n').slice(2)

    let totalRx = 0
    let totalTx = 0

    for (const line of lines) {
      const parts = line.trim().split(/\s+/)
      if (parts.length >= 10 && !parts[0].startsWith('lo:')) {
        totalRx += parseInt(parts[1], 10) || 0
        totalTx += parseInt(parts[9], 10) || 0
      }
    }

    return { rx: totalRx, tx: totalTx }
  } catch {
    return undefined
  }
}

function getDiskStats(): { read: number; write: number } | undefined {
  if (process.platform !== 'linux') {
    return undefined
  }

  try {
    const diskStats = fs.readFileSync('/proc/diskstats', 'utf-8')
    const lines = diskStats.split('\n')

    let totalRead = 0
    let totalWrite = 0

    for (const line of lines) {
      const parts = line.trim().split(/\s+/)
      if (parts.length >= 14) {
        totalRead += (parseInt(parts[5], 10) || 0) * 512
        totalWrite += (parseInt(parts[9], 10) || 0) * 512
      }
    }

    return { read: totalRead, write: totalWrite }
  } catch {
    return undefined
  }
}

function getOpenFileDescriptors(): number | undefined {
  if (process.platform === 'linux') {
    try {
      const fdDir = `/proc/${process.pid}/fd`
      const fds = fs.readdirSync(fdDir)
      return fds.length
    } catch {
      return undefined
    }
  }

  return undefined
}

export function collectMetrics(): WorkerMetrics {
  const now = Date.now()
  const uptimeSecs = Math.floor((now - processStartTime) / 1000)

  const windowDurationSecs = (now - lastInvocationWindow) / 1000
  const invocationsPerSec = windowDurationSecs > 0 ? invocationCount / windowDurationSecs : 0
  const avgLatencyMs = invocationCount > 0 ? totalLatencyMs / invocationCount : 0

  let networkRx: number | undefined
  let networkTx: number | undefined
  let diskRead: number | undefined
  let diskWrite: number | undefined

  const currentNetStats = getNetworkStats()
  if (currentNetStats && lastNetworkStats) {
    networkRx = currentNetStats.rx - lastNetworkStats.rx
    networkTx = currentNetStats.tx - lastNetworkStats.tx
  }
  if (currentNetStats) {
    lastNetworkStats = { ...currentNetStats, time: now }
  }

  const currentDiskStats = getDiskStats()
  if (currentDiskStats && lastDiskStats) {
    diskRead = currentDiskStats.read - lastDiskStats.read
    diskWrite = currentDiskStats.write - lastDiskStats.write
  }
  if (currentDiskStats) {
    lastDiskStats = { ...currentDiskStats, time: now }
  }

  const memMetrics = getMemoryMetrics()

  const process_metrics: ProcessMetrics = {
    cpu_percent: getCpuPercent(),
    memory_used_bytes: memMetrics.used,
    memory_total_bytes: memMetrics.total,
    process_uptime_secs: uptimeSecs,
  }

  const performance: PerformanceMetrics = {
    thread_count: getThreadCount(),
    invocations_per_sec: invocationsPerSec,
    avg_latency_ms: avgLatencyMs,
  }

  const extended: ExtendedMetrics = {
    disk_read_bytes: diskRead,
    disk_write_bytes: diskWrite,
    network_rx_bytes: networkRx,
    network_tx_bytes: networkTx,
    open_file_descriptors: getOpenFileDescriptors(),
    error_count: errorCount,
  }

  invocationCount = 0
  totalLatencyMs = 0
  errorCount = 0
  lastInvocationWindow = now

  const metrics: WorkerMetrics = {
    collected_at_ms: now,
    process: process_metrics,
    performance,
    extended,
  }

  if (isKubernetes()) {
    metrics.k8s_identifiers = getKubernetesIdentifiers()
    metrics.k8s_core = getKubernetesCoreMetrics()
    metrics.k8s_resources = getKubernetesResourceMetrics()
    metrics.k8s_extended = getKubernetesExtendedMetrics()
  }

  return metrics
}

export function createMetricsReporter(
  reportFn: (metrics: Omit<WorkerMetrics, 'collected_at_ms'>) => void,
  intervalMs: number = 5000
): { start: () => void; stop: () => void } {
  let timer: NodeJS.Timeout | null = null

  return {
    start: () => {
      if (timer) return

      collectMetrics()

      timer = setInterval(() => {
        const metrics = collectMetrics()
        reportFn(metrics)
      }, intervalMs)
    },
    stop: () => {
      if (timer) {
        clearInterval(timer)
        timer = null
      }
    },
  }
}

File: packages/rust/iii/src/metrics.rs
use std::fs;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Mutex;
use std::time::Instant;

use sysinfo::{Disks, Networks, Pid, System};

use crate::bridge::{
    ExtendedMetrics, KubernetesCoreMetrics, KubernetesExtendedMetrics, KubernetesIdentifiers,
    KubernetesResourceMetrics, PerformanceMetrics, ProcessMetrics, WorkerMetrics,
};

static INVOCATION_COUNT: AtomicU64 = AtomicU64::new(0);
static TOTAL_LATENCY_MS: AtomicU64 = AtomicU64::new(0);
static ERROR_COUNT: AtomicU64 = AtomicU64::new(0);

lazy_static::lazy_static! {
    static ref START_TIME: Instant = Instant::now();
    static ref SYSTEM: Mutex<System> = Mutex::new(System::new_all());
    static ref NETWORKS: Mutex<Networks> = Mutex::new(Networks::new_with_refreshed_list());
    static ref DISKS: Mutex<Disks> = Mutex::new(Disks::new_with_refreshed_list());
    static ref LAST_NETWORK_STATS: Mutex<Option<(u64, u64)>> = Mutex::new(None);
    static ref LAST_DISK_STATS: Mutex<Option<(u64, u64)>> = Mutex::new(None);
}

pub fn record_invocation(latency_ms: u64, is_error: bool) {
    INVOCATION_COUNT.fetch_add(1, Ordering::Relaxed);
    TOTAL_LATENCY_MS.fetch_add(latency_ms, Ordering::Relaxed);
    if is_error {
        ERROR_COUNT.fetch_add(1, Ordering::Relaxed);
    }
}

pub fn is_kubernetes() -> bool {
    std::env::var("KUBERNETES_SERVICE_HOST").is_ok()
}

fn get_kubernetes_identifiers() -> Option<KubernetesIdentifiers> {
    if !is_kubernetes() {
        return None;
    }

    Some(KubernetesIdentifiers {
        cluster: std::env::var("CLUSTER_NAME").ok(),
        namespace: std::env::var("POD_NAMESPACE")
            .ok()
            .or_else(|| read_k8s_file("namespace")),
        pod_name: std::env::var("POD_NAME")
            .ok()
            .or_else(|| std::env::var("HOSTNAME").ok()),
        container_name: std::env::var("CONTAINER_NAME").ok(),
        node_name: std::env::var("NODE_NAME").ok(),
        pod_uid: std::env::var("POD_UID")
            .ok()
            .or_else(|| read_k8s_file("pod-uid")),
    })
}

fn read_k8s_file(filename: &str) -> Option<String> {
    let path = format!("/var/run/secrets/kubernetes.io/serviceaccount/{}", filename);
    std::fs::read_to_string(path).ok().map(|s| s.trim().to_string())
}

fn read_cgroup_value(v1_path: &str, v2_path: &str) -> Option<u64> {
    for path in &[v2_path, v1_path] {
        if let Ok(content) = fs::read_to_string(path) {
            if let Ok(value) = content.trim().parse::<u64>() {
                if value != u64::MAX {
                    return Some(value);
                }
            }
        }
    }
    None
}

fn get_cgroup_cpu_usage_cores() -> Option<f64> {
    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/cpu.stat") {
        for line in content.lines() {
            if line.starts_with("usage_usec") {
                if let Some(usec_str) = line.split_whitespace().nth(1) {
                    if let Ok(usec) = usec_str.parse::<u64>() {
                        return Some(usec as f64 / 1_000_000.0);
                    }
                }
            }
        }
    }

    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/cpu/cpuacct.usage") {
        if let Ok(nsec) = content.trim().parse::<u64>() {
            return Some(nsec as f64 / 1_000_000_000.0);
        }
    }

    None
}

fn get_cgroup_memory_working_set() -> Option<u64> {
    read_cgroup_value(
        "/sys/fs/cgroup/memory/memory.usage_in_bytes",
        "/sys/fs/cgroup/memory.current",
    )
}

fn get_cgroup_memory_limit() -> Option<u64> {
    let value = read_cgroup_value(
        "/sys/fs/cgroup/memory/memory.limit_in_bytes",
        "/sys/fs/cgroup/memory.max",
    )?;

    if value < 9223372036854771712 {
        Some(value)
    } else {
        None
    }
}

fn get_cgroup_cpu_limit_cores() -> Option<f64> {
    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/cpu.max") {
        let content = content.trim();
        if content != "max" {
            let parts: Vec<&str> = content.split_whitespace().collect();
            if parts.len() >= 2 {
                if let (Ok(quota), Ok(period)) = (parts[0].parse::<i64>(), parts[1].parse::<i64>())
                {
                    if quota > 0 && period > 0 {
                        return Some(quota as f64 / period as f64);
                    }
                }
            }
        }
    }

    if let (Ok(quota_str), Ok(period_str)) = (
        fs::read_to_string("/sys/fs/cgroup/cpu/cpu.cfs_quota_us"),
        fs::read_to_string("/sys/fs/cgroup/cpu/cpu.cfs_period_us"),
    ) {
        if let (Ok(quota), Ok(period)) = (
            quota_str.trim().parse::<i64>(),
            period_str.trim().parse::<i64>(),
        ) {
            if quota > 0 && period > 0 {
                return Some(quota as f64 / period as f64);
            }
        }
    }

    None
}

fn get_cgroup_cpu_throttled_seconds() -> Option<f64> {
    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/cpu.stat") {
        for line in content.lines() {
            if line.starts_with("throttled_usec") {
                if let Some(usec_str) = line.split_whitespace().nth(1) {
                    if let Ok(usec) = usec_str.parse::<u64>() {
                        return Some(usec as f64 / 1_000_000.0);
                    }
                }
            }
        }
    }

    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/cpu/cpu.stat") {
        for line in content.lines() {
            if line.starts_with("throttled_time") {
                if let Some(ns_str) = line.split_whitespace().nth(1) {
                    if let Ok(ns) = ns_str.parse::<u64>() {
                        return Some(ns as f64 / 1_000_000_000.0);
                    }
                }
            }
        }
    }

    None
}

fn parse_k8s_memory(value: &str) -> Option<u64> {
    let value = value.trim();

    let (num_str, unit) = if let Some(pos) = value.find(|c: char| c.is_alphabetic()) {
        (&value[..pos], &value[pos..])
    } else {
        (value, "")
    };

    let num: f64 = num_str.parse().ok()?;

    let multiplier = match unit {
        "" => 1,
        "K" => 1000,
        "M" => 1000 * 1000,
        "G" => 1000 * 1000 * 1000,
        "T" => 1000u64.pow(4),
        "P" => 1000u64.pow(5),
        "E" => 1000u64.pow(6),
        "Ki" => 1024,
        "Mi" => 1024 * 1024,
        "Gi" => 1024 * 1024 * 1024,
        "Ti" => 1024u64.pow(4),
        "Pi" => 1024u64.pow(5),
        "Ei" => 1024u64.pow(6),
        _ => return None,
    };

    Some((num * multiplier as f64) as u64)
}

#[cfg(target_os = "linux")]
fn get_disk_io_stats() -> Option<(u64, u64)> {
    let content = fs::read_to_string("/proc/diskstats").ok()?;

    let mut total_read = 0u64;
    let mut total_write = 0u64;

    for line in content.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 14 {
            if let (Ok(read), Ok(write)) = (parts[5].parse::<u64>(), parts[9].parse::<u64>()) {
                total_read += read * 512;
                total_write += write * 512;
            }
        }
    }

    Some((total_read, total_write))
}

#[cfg(not(target_os = "linux"))]
fn get_disk_io_stats() -> Option<(u64, u64)> {
    None
}

#[cfg(target_os = "linux")]
fn get_open_file_descriptors() -> Option<u32> {
    let fd_path = format!("/proc/{}/fd", std::process::id());
    match fs::read_dir(&fd_path) {
        Ok(entries) => Some(entries.count() as u32),
        Err(_) => None,
    }
}

#[cfg(not(target_os = "linux"))]
fn get_open_file_descriptors() -> Option<u32> {
    None
}

fn get_filesystem_usage() -> Option<u64> {
    #[cfg(target_os = "linux")]
    {
        if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/io.stat") {
            for line in content.lines() {
                if line.contains("wbytes=") || line.contains("rbytes=") {
                }
            }
        }

        use std::ffi::CString;
        use std::mem::MaybeUninit;

        let path = CString::new("/").ok()?;
        let mut stat: MaybeUninit<libc::statvfs> = MaybeUninit::uninit();

        unsafe {
            if libc::statvfs(path.as_ptr(), stat.as_mut_ptr()) == 0 {
                let stat = stat.assume_init();
                let block_size = stat.f_frsize as u64;
                let total_blocks = stat.f_blocks as u64;
                let free_blocks = stat.f_bfree as u64;
                let used = (total_blocks - free_blocks) * block_size;
                return Some(used);
            }
        }
    }

    None
}

fn get_node_pressure() -> (Option<bool>, Option<bool>, Option<bool>) {
    let mut memory_pressure = None;
    let mut disk_pressure = None;
    let mut pid_pressure = None;

    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/memory.pressure") {
        for line in content.lines() {
            if line.starts_with("some") {
                if let Some(avg10) = extract_pressure_avg10(&line) {
                    memory_pressure = Some(avg10 > 10.0);
                }
            }
        }
    }

    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/io.pressure") {
        for line in content.lines() {
            if line.starts_with("some") {
                if let Some(avg10) = extract_pressure_avg10(&line) {
                    disk_pressure = Some(avg10 > 10.0);
                }
            }
        }
    }

    if let Ok(content) = fs::read_to_string("/sys/fs/cgroup/cpu.pressure") {
        for line in content.lines() {
            if line.starts_with("some") {
                if let Some(avg10) = extract_pressure_avg10(&line) {
                    pid_pressure = Some(avg10 > 10.0);
                }
            }
        }
    }

    (memory_pressure, disk_pressure, pid_pressure)
}

fn extract_pressure_avg10(line: &str) -> Option<f64> {
    for part in line.split_whitespace() {
        if part.starts_with("avg10=") {
            return part.trim_start_matches("avg10=").parse().ok();
        }
    }
    None
}

fn get_kubernetes_resource_metrics() -> Option<KubernetesResourceMetrics> {
    if !is_kubernetes() {
        return None;
    }

    let mut metrics = KubernetesResourceMetrics::default();

    metrics.cpu_limits_cores = get_cgroup_cpu_limit_cores();
    metrics.memory_limits_bytes = get_cgroup_memory_limit();
    metrics.cpu_throttled_seconds_total = get_cgroup_cpu_throttled_seconds();

    if let Ok(cpu_request) = std::env::var("CPU_REQUEST") {
        if cpu_request.ends_with('m') {
            if let Ok(millicores) = cpu_request.trim_end_matches('m').parse::<u64>() {
                metrics.cpu_requests_cores = Some(millicores as f64 / 1000.0);
            }
        } else if let Ok(cores) = cpu_request.parse::<f64>() {
            metrics.cpu_requests_cores = Some(cores);
        }
    }

    if let Ok(mem_request) = std::env::var("MEMORY_REQUEST") {
        metrics.memory_requests_bytes = parse_k8s_memory(&mem_request);
    }

    Some(metrics)
}

fn get_kubernetes_extended_metrics() -> Option<KubernetesExtendedMetrics> {
    if !is_kubernetes() {
        return None;
    }

    let mut metrics = KubernetesExtendedMetrics::default();

    let networks = NETWORKS.lock().unwrap();
    let mut rx = 0u64;
    let mut tx = 0u64;
    for (_name, data) in networks.iter() {
        rx += data.total_received();
        tx += data.total_transmitted();
    }
    metrics.network_rx_bytes_total = Some(rx);
    metrics.network_tx_bytes_total = Some(tx);

    metrics.fs_usage_bytes = get_filesystem_usage();

    let (mem_pressure, disk_pressure, pid_pressure) = get_node_pressure();
    metrics.node_memory_pressure = mem_pressure;
    metrics.node_disk_pressure = disk_pressure;
    metrics.node_pid_pressure = pid_pressure;

    Some(metrics)
}

fn get_kubernetes_core_metrics() -> Option<KubernetesCoreMetrics> {
    if !is_kubernetes() {
        return None;
    }

    Some(KubernetesCoreMetrics {
        uptime_seconds: Some(START_TIME.elapsed().as_secs()),
        pod_phase: Some("Running".to_string()),
        pod_ready: Some(true),
        cpu_usage_cores: get_cgroup_cpu_usage_cores(),
        memory_working_set_bytes: get_cgroup_memory_working_set(),
        container_restarts_total: std::env::var("CONTAINER_RESTARTS")
            .ok()
            .and_then(|s| s.parse().ok()),
        last_termination_reason: std::env::var("LAST_TERMINATION_REASON").ok(),
    })
}

pub fn collect_metrics() -> WorkerMetrics {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis() as u64;

    let uptime_secs = START_TIME.elapsed().as_secs();

    let (cpu_percent, memory_used, memory_total, thread_count) = {
        let mut sys = SYSTEM.lock().unwrap();
        sys.refresh_all();

        let pid = Pid::from_u32(std::process::id());

        if let Some(process) = sys.process(pid) {
            (
                process.cpu_usage() as f64,
                process.memory(),
                sys.total_memory(),
                1u32,
            )
        } else {
            (0.0, 0, sys.total_memory(), 1)
        }
    };

    let (network_rx, network_tx) = {
        let mut networks = NETWORKS.lock().unwrap();
        networks.refresh();

        let mut rx = 0u64;
        let mut tx = 0u64;
        for (_name, data) in networks.iter() {
            rx += data.total_received();
            tx += data.total_transmitted();
        }

        let mut last_stats = LAST_NETWORK_STATS.lock().unwrap();
        let delta = if let Some((last_rx, last_tx)) = *last_stats {
            (
                rx.saturating_sub(last_rx),
                tx.saturating_sub(last_tx),
            )
        } else {
            (0, 0)
        };
        *last_stats = Some((rx, tx));
        delta
    };

    let (disk_read, disk_write) = {
        if let Some((current_read, current_write)) = get_disk_io_stats() {
            let mut last_stats = LAST_DISK_STATS.lock().unwrap();
            let delta = if let Some((last_read, last_write)) = *last_stats {
                (
                    current_read.saturating_sub(last_read),
                    current_write.saturating_sub(last_write),
                )
            } else {
                (0, 0)
            };
            *last_stats = Some((current_read, current_write));
            (Some(delta.0), Some(delta.1))
        } else {
            (None, None)
        }
    };

    let open_fds = get_open_file_descriptors();

    let invocation_count = INVOCATION_COUNT.swap(0, Ordering::Relaxed);
    let total_latency = TOTAL_LATENCY_MS.swap(0, Ordering::Relaxed);
    let error_count = ERROR_COUNT.load(Ordering::Relaxed);

    let invocations_per_sec = invocation_count as f64 / 5.0;
    let avg_latency_ms = if invocation_count > 0 {
        total_latency as f64 / invocation_count as f64
    } else {
        0.0
    };

    let process = ProcessMetrics {
        cpu_percent: Some(cpu_percent),
        memory_used_bytes: Some(memory_used),
        memory_total_bytes: Some(memory_total),
        process_uptime_secs: Some(uptime_secs),
    };

    let performance = PerformanceMetrics {
        thread_count: Some(thread_count),
        open_connections: Some(1),
        invocations_per_sec: Some(invocations_per_sec),
        avg_latency_ms: Some(avg_latency_ms),
    };

    let extended = ExtendedMetrics {
        disk_read_bytes: disk_read,
        disk_write_bytes: disk_write,
        network_rx_bytes: Some(network_rx),
        network_tx_bytes: Some(network_tx),
        open_file_descriptors: open_fds,
        error_count: Some(error_count),
    };

    let mut metrics = WorkerMetrics {
        collected_at_ms: now,
        process,
        performance,
        extended,
        ..Default::default()
    };

    if is_kubernetes() {
        metrics.k8s_identifiers = get_kubernetes_identifiers();
        metrics.k8s_core = get_kubernetes_core_metrics();
        metrics.k8s_resources = get_kubernetes_resource_metrics();
        metrics.k8s_extended = get_kubernetes_extended_metrics();
    }

    metrics
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_collect_metrics() {
        let metrics = collect_metrics();
        assert!(metrics.collected_at_ms > 0);
        assert!(metrics.process.process_uptime_secs.is_some());
    }

    #[test]
    fn test_record_invocation() {
        record_invocation(100, false);
        record_invocation(200, true);

        let metrics = collect_metrics();
        assert!(metrics.performance.avg_latency_ms.is_some());
    }
}
